cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -1213,10 +1218,125 @@ Status NumPyConverter::ConvertLists(const std::shared_ptr<DataType>& type) {
   return PushBuilderResult(list_builder);
 }
 
+Status NumPyConverter::Visit(const BinaryType& type) {
+  BinaryBuilder builder(pool_);
+
+  auto data = reinterpret_cast<const uint8_t*>(PyArray_DATA(arr_));
+
+  int item_length = 0;
+  if (mask_ != nullptr) {
+    Ndarray1DIndexer<uint8_t> mask_values(mask_);
+    for (int64_t i = 0; i < length_; ++i) {
+      if (mask_values[i]) {
+        RETURN_NOT_OK(builder.AppendNull());
+      } else {
+        // This is annoying. NumPy allows strings to have nul-terminators, so
+        // we must check for them here
+        for (item_length = 0; item_length < itemsize_; ++item_length) {
+          if (data[item_length] == 0) {
```

Comment:
I think this can be made a bit cleaner by using `std::find`.

```cpp
builder.Append(data, std::distance(data, std::find(data, data + itemsize_, '\0')))
```

Reply:
Ah, guess I'm a C programmer at heart